dbt/include/athena/macros/materializations/models/table/create_table_as.sql

```diff
@@ -22,6 +22,10 @@
                                                relation.s3_path_table_part,
                                                external_location,
                                                temporary) -%}
+  {%- set contract_config = config.get('contract') -%}
+  {%- if contract_config.enforced -%}
+    {{ get_assert_columns_equivalent(sql) }}
+  {%- endif -%}
```

Comment:
as the contract check was added in the create_table_as does it mean that also `table_hive_ha` is supported as it uses `create_table_as` ? if so could you double check that?

Reply:
This means that the contract will also be applied to temporary tables. We have to check it does not lead to issues, especially in the `delete_condition` for Iceberg because for this, it might depend on a table which is the source table (the temp table) and not the destination table (the incremental table).

Reply:
Make sense, and it's ideal, when we built `table_hive_ha` we wanted to use as much possible building blocks from the adapter.

https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/table/table_hive_ha.sql#L35

https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/table/table_hive_ha.sql#L47

Let's just try to see if some work is required. - e.g. add the contracts for a table_hive_ha and see if it works.

Reply:
@Jrmyy yeah let's try out for iceberg too. 

I don't fully get the implementation works in general,  why the get_assert_columns_equivalent is called before we run the cta and not after?

Reply:
In a nutshell ([here](url) for the full explanation)
- dbt is going to create two queries to compare the schemas, one query based on the SQL and the other query based on the YAML
- To create the SQL query based on the SQL provided by the models it uses this [macro](https://github.com/dbt-labs/dbt-core/blob/f1dddaa6e96b743498440925a55e66a7d5cdacfa/core/dbt/include/global_project/macros/adapters/columns.sql#L56). The advantage of this is that no data scan so no money wasted 💸 
- To create the SQL query based on the YAML, it uses the same macro as before, but first it needs to create the query. And it uses this [macro](https://github.com/dbt-labs/dbt-core/blob/f1dddaa6e96b743498440925a55e66a7d5cdacfa/core/dbt/include/global_project/macros/adapters/columns.sql#L41)
- After that this is just string comparisons

The main advantage of this method is that it does real type checking on the database system so that we are sure that the types defined in the contract match real types, understood by the database system (in our case Athena).

Why do I do this before ?
To prevent from table creation and therefore money spent wrongly on data scanning because model will be failed since the contract is enforced and there are type issues.

Reply:
thanks for the awsome explanation, now is clear https://github.com/dbt-labs/dbt-core/blob/f1dddaa6e96b743498440925a55e66a7d5cdacfa/core/dbt/include/global_project/macros/adapters/columns.sql#L56 thanks to this they create the SQL table, and compare then with the types specified in the contracts.

Reply:
In fact we cannot use contract for `table_hive_ha` 😞 
https://github.com/dbt-labs/dbt-core/blob/f1dddaa6e96b743498440925a55e66a7d5cdacfa/core/dbt/parser/schemas.py#L1025

Reply:
We might want to incorporate at some point the `table_hive_ha` to `table` materialization with the right config.

Reply:
thanks for spotting this, I'm in favor of moving table_hive_ha to table, and setting a flag to use explicitly `ha` (e.g. ha=true).

Reply:
Contracts works with Iceberg since it is in `table` or `incremental` materialization.